zip.go

```diff
@@ -317,6 +328,18 @@ func (z *Zip) Write(f File) error {
 		return nil
 	}
 
+	if (header.Mode() & os.ModeSymlink) != 0 {
```

Comment:
What do you think about mirroring the more elegant switch over in the tar extractor? (perhaps sans hard link of course):

https://sourcegraph.com/github.com/mholt/archiver@d572b2e8b82726cee9476d1b9d63a7fe9b601ff1/-/blob/tar.go#L232

Reply:
I don't understand how this should work. The TAR `Header` has an explicit `Typeflag` byte, that you can switch on:

https://github.com/golang/go/blob/ea509c9/src/archive/tar/common.go#L144

The ZIP `FileHeader` has a `Mode()` function that needs to locate the mode bits depending on the "creator" version:

https://github.com/golang/go/blob/ea509c9/src/archive/zip/struct.go#L277-L289

So we would have to switch on `header.Mode() & os.ModeType` and could then check for `os.ModeSymLink`, and an explicit `0` for `IsRegular()`? Essentially:

```
	switch header.Mode() & os.ModeType {
	case os.ModeSymlink:
		linkTarget, err := os.Readlink(f.Name())
		if err != nil {
			return fmt.Errorf("%s: readlink: %v", f.Name(), err)
		}
		_, err = writer.Write([]byte(filepath.ToSlash(linkTarget)))
		if err != nil {
			return fmt.Errorf("%s: writing symlink target: %v", f.Name(), err)
		}
	case 0: // regular file
		if f.ReadCloser == nil {
			return fmt.Errorf("%s: no way to read file contents", f.Name())
		}
		_, err := io.Copy(writer, f)
		if err != nil {
			return fmt.Errorf("%s: copying contents: %v", f.Name(), err)
		}
	}
	return nil
}
```

I can do this if you prefer; let me know.

Reply:
Oh, I see what you mean now.

Hmm, maybe a switch isn't quite right then, you've got a point. An `if` statement could do. Then perhaps this should be brought out into its own method, like `z.writeFile`. That'll satisfy my OCD on this. :smile:

Reply:
I don't know; moving it into a separate function will require duplication of the `header` assertion logic, so if there is only a single caller, I would prefer to avoid the code duplication.

I've pushed the changes I have so far, as I'm done for today. I can always rewrite it some more if you don't like the switch either...

Reply:
> That'll satisfy my OCD on this.

Talking about OCD, it took me a lot of willpower not to change `hdr` to `header` in the rest of the codebase to get consistent naming everywhere. 😄

Reply:
Heh, you shouldn't have told me about that 😣